dataplane/src/nat.rs

```diff
@@ -19,88 +27,60 @@ struct Pif {
     vpc: String,                  // Name of the associated VPC
 }
 
-#[derive(Default, Debug, Clone)]
-struct TrieNode<K, V>
-where
-    K: Clone + Eq + Hash,
-    V: Clone,
-{
-    children: HashMap<K, TrieNode<K, V>>,
-    value: Option<V>,
-}
-
-#[derive(Default, Debug, Clone)]
-struct PrefixTrie<K, V>
-where
-    K: Clone + Eq + Hash,
-    V: Clone,
-{
-    root: TrieNode<K, V>,
+#[derive(Default, Clone)]
+struct PrefixTrie {
+    trie_ipv4: RTrieMap<Ipv4Prefix, String>,
+    trie_ipv6: RTrieMap<Ipv6Prefix, String>,
 }
 
-impl<K, V> PrefixTrie<K, V>
-where
-    K: Clone + Eq + Hash + Default + Debug,
-    V: Clone + Default + Debug,
-{
+impl PrefixTrie {
     #[tracing::instrument(level = "trace")]
     fn new() -> Self {
         Self {
-            root: TrieNode::default(),
+            trie_ipv4: RTrieMap::new(),
+            trie_ipv6: RTrieMap::new(),
         }
     }
 
     #[tracing::instrument(level = "trace")]
-    fn insert(&mut self, keys: Vec<K>, value: V) -> bool {
-        let mut node = &mut self.root;
-
-        for key in keys {
-            node = node.children.entry(key).or_default();
-        }
-
-        if node.value.is_some() {
-            return false; // Prefix already exists
-        }
-
-        node.value = Some(value);
-        true
-    }
-
-    #[tracing::instrument(level = "trace")]
-    fn find(&self, keys: Vec<K>) -> Option<V> {
-        let mut node = &self.root;
-        let mut best_match = None;
-
-        for key in keys {
-            if let Some(val) = &node.value {
-                best_match = Some(val.clone());
+    fn insert(&mut self, addr: &IpAddr, len: &u8, value: String) -> Result<(), NatError> {
```

Comment:
This is absolutely fine. However, in https://github.com/githedgehog/dataplane/pull/211, I added an enum Prefix that works for both IPv4 and IPv6. I would use that type here so that:
   * the same implementation works also for IPv6 (with minor tweaks like using trie_ipv4 or trie_ipv6)
   * we don't leak the Ipv4Prefix, Ipv6Prefix type (specific to the iptrie) outside of the "scope" where they are absolutely needed. The Prefix uses "standard" types.
   * we have a common type to represent prefixes throughout.

Reply:
Oh, sorry, now I realize that you added also IPv6.
Up to you then to use Prefix.

Reply:
I'll take a look, thanks.

Note also that iptrie [might support both IPv4 and IPv6 in a single map in the near future](https://github.com/Orange-OpenSource/iptrie/issues/3#issuecomment-2658574297).